SimPEG/EM/Utils/CurrentUtils.py

```diff
@@ -0,0 +1,228 @@
+import numpy as np
+from SimPEG import Utils
+
+
+def line(a, t, l):
+    """
+        Linear interpolation between a and b
+        0 <= t <= 1
+    """
+    return a + t * l
+
+
+def weight(t, a1, l1, h1, a2, l2, h2):
+    """
+        Edge basis functions
+    """
+    x1 = line(a1, t, l1)
+    x2 = line(a2, t, l2)
+    w0 = (1. - x1 / h1) * (1. - x2 / h2)
+    w1 = (x1 / h1) * (1. - x2 / h2)
+    w2 = (1. - x1 / h1) * (x2 / h2)
+    w3 = (x1 / h1) * (x2 / h2)
+    return np.r_[w0, w1, w2, w3]
+
+
+# TODO: Extend this when current is defined on cell-face
+def getStraightLineCurrentIntegral(hx, hy, hz, ax, ay, az, bx, by, bz):
+    """
+      Compute integral int(W . J dx^3) in brick of size hx x hy x hz
+      where W denotes the 12 local bilinear edge basis functions
+      and where J prescribes a unit line current
+      between points (ax,ay,az) and (bx,by,bz).
+    """
+
+    # length of line segment
+    lx = bx - ax
+    ly = by - ay
+    lz = bz - az
+    l = np.sqrt(lx**2+ly**2+lz**2)
+
+    if l == 0:
+        sx = np.zeros(4, 1)
+        sy = np.zeros(4, 1)
+        sz = np.zeros(4, 1)
+
+    # integration using Simpson's rule
+    wx0 = weight(0., ay, ly, hy, az, lz, hz)
+    wx0_5 = weight(0.5, ay, ly, hy, az, lz, hz)
+    wx1 = weight(1., ay, ly, hy, az, lz, hz)
+
+    wy0 = weight(0., ax, lx, hx, az, lz, hz)
+    wy0_5 = weight(0.5, ax, lx, hx, az, lz, hz)
+    wy1 = weight(1., ax, lx, hx, az, lz, hz)
+
+    wz0 = weight(0., ax, lx, hx, ay, ly, hy)
+    wz0_5 = weight(0.5, ax, lx, hx, ay, ly, hy)
+    wz1 = weight(1., ax, lx, hx, ay, ly, hy)
+
+    sx = (wx0 + 4. * wx0_5 + wx1) * (lx / 6.)
+
+    sy = (wy0 + 4. * wy0_5 + wy1) * (ly / 6.)
+    sz = (wz0 + 4. * wz0_5 + wz1) * (lz / 6.)
+
+    return sx, sy, sz
+
+
+def findlast(x):
+    if x.sum() == 0:
+        return -1
+    else:
+        return np.arange(x.size)[x][-1]
+
+
+def getSourceTermLineCurrentPolygon(xorig, hx, hy, hz, px, py, pz):
+    """
+        Given a tensor product mesh with origin at (x0,y0,z0) and cell sizes
+        hx, hy, hz, compute the source vector for a unit current flowing along
+        the polygon with vertices px, py, pz.
+        The 3-D arrays sx, sy, sz contain the source terms for all x/y/z-edges
+        of the tensor product mesh.
+
+        Modified from matlab code:
```

Comment:
@sgkang: can you please confirm that Christoph is ok with this? this is an open source implementation of originally closed-source code. If his implementation is already in jinv, then we can reference that, or a paper of his.